cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/release.go

```diff
@@ -353,23 +363,73 @@ func (s *Server) newRelease(ctx context.Context, packageRef *corev1.AvailablePac
 	return &corev1.InstalledPackageReference{
 		Context:    &corev1.Context{Namespace: name.Namespace},
 		Identifier: name.Name,
+		Plugin:     GetPluginDetail(),
 	}, nil
 }
 
+// returns 3 things:
+// - ready:  whether the HelmRelease object is in a ready state
+// - reason: one of SUCCESS/FAILURE/PENDING/UNKNOWN, if present
+// - userReason: textual description of why the object is in current state, if present
+// docs:
+// 1. https://fluxcd.io/docs/components/helm/helmreleases/#examples
+// 2. discussion on https://vmware.slack.com/archives/C4HEXCX3N/p1630907107078800.
+func isHelmReleaseReady(unstructuredObj map[string]interface{}) (ready bool, status corev1.InstalledPackageStatus_StatusReason, userReason string) {
+	if !checkGeneration(unstructuredObj) {
+		return false, corev1.InstalledPackageStatus_STATUS_REASON_UNSPECIFIED, ""
+	}
+
+	conditions, found, err := unstructured.NestedSlice(unstructuredObj, "status", "conditions")
+	if err != nil || !found {
+		return false, corev1.InstalledPackageStatus_STATUS_REASON_UNSPECIFIED, ""
+	}
+
+	isInstallFailed := false
+
+	for _, conditionUnstructured := range conditions {
+		if conditionAsMap, ok := conditionUnstructured.(map[string]interface{}); ok {
+			if typeString, ok := conditionAsMap["type"]; ok && typeString == "Ready" {
+				// this could be something like
+				// "reason": "InstallFailed"
+				// i.e. not super-useful
+				if reasonString, ok := conditionAsMap["reason"]; ok {
+					userReason = fmt.Sprintf("%v", reasonString)
+					if reasonString == "InstallFailed" {
+						isInstallFailed = true
+					}
+				}
+				// whereas this could be something like:
+				// "message": 'Helm install failed: unable to build kubernetes objects from
+				// release manifest: error validating "": error validating data:
+				// ValidationError(Deployment.spec.replicas): invalid type for
+				// io.k8s.api.apps.v1.DeploymentSpec.replicas: got "string", expected "integer"'
+				// i.e. a little more useful, so we'll just return them both
+				if messageString, ok := conditionAsMap["message"]; ok {
+					userReason += fmt.Sprintf(": %v", messageString)
+				}
+				if statusString, ok := conditionAsMap["status"]; ok {
+					if statusString == "True" {
+						return true, corev1.InstalledPackageStatus_STATUS_REASON_INSTALLED, userReason
```

Comment:
not theoretical, real case. tested and reproducible, if needed. But i will move the check above as you asked.

Reply:
just fyi, here is a real-life example from an integration test I am working on that shows this use case:
```
[{
  "apiVersion": "helm.toolkit.fluxcd.io/v2beta1",
  "kind": "HelmRelease",
  "metadata": {
    "creationTimestamp": "2021-09-14T03:27:41Z",
    "finalizers": [
      "finalizers.fluxcd.io"
    ],
    "generation": 2,
    "managedFields": [
      {
        "apiVersion": "helm.toolkit.fluxcd.io/v2beta1",
        "fieldsType": "FieldsV1",
        "fieldsV1": {
          "f:spec": {
            ".": {},
            "f:chart": {
              ".": {},
              "f:spec": {
                ".": {},
                "f:chart": {},
                "f:sourceRef": {
                  ".": {},
                  "f:kind": {},
                  "f:name": {},
                  "f:namespace": {}
                },
                "f:version": {}
              }
            },
            "f:install": {
              ".": {},
              "f:createNamespace": {}
            },
            "f:targetNamespace": {},
            "f:values": {}
          }
        },
        "manager": "kubeapps-apis",
        "operation": "Update",
        "time": "2021-09-14T03:27:41Z"
      },
      {
        "apiVersion": "helm.toolkit.fluxcd.io/v2beta1",
        "fieldsType": "FieldsV1",
        "fieldsV1": {
          "f:metadata": {
            "f:finalizers": {
              ".": {},
              "v:\"finalizers.fluxcd.io\"": {}
            }
          },
          "f:spec": {
            "f:interval": {}
          },
          "f:status": {
            "f:conditions": {},
            "f:failures": {},
            "f:helmChart": {},
            "f:installFailures": {},
            "f:lastAttemptedRevision": {},
            "f:lastAttemptedValuesChecksum": {},
            "f:observedGeneration": {}
          }
        },
        "manager": "helm-controller",
        "operation": "Update",
        "time": "2021-09-14T03:27:42Z"
      }
    ],
    "name": "my-podinfo-5",
    "namespace": "kubeapps",
    "resourceVersion": "1707794",
    "uid": "9e574bd3-61bd-4b5f-9b46-33466c4b8773"
  },
  "spec": {
    "chart": {
      "spec": {
        "chart": "podinfo",
        "sourceRef": {
          "kind": "HelmRepository",
          "name": "podinfo-5",
          "namespace": "default"
        },
        "version": "*"
      }
    },
    "install": {
      "createNamespace": true
    },
    "interval": "1m0s",
    "targetNamespace": "test-5-mdnt",
    "values": {
      "replicaCount": "what we do in the shadows"
    }
  },
  "status": {
    "conditions": [
      {
        "lastTransitionTime": "2021-09-14T03:27:42Z",
        "message": "install retries exhausted",
        "reason": "InstallFailed",
        "status": "False",
        "type": "Ready"
      },
      {
        "lastTransitionTime": "2021-09-14T03:27:42Z",
        "message": "Helm install failed: unable to build kubernetes objects from release manifest: error validating \"\": error validating data: ValidationError(Deployment.spec.replicas): invalid type for io.k8s.api.apps.v1.DeploymentSpec.replicas: got \"string\", expected \"integer\"\n\nLast Helm logs:\n\n",
        "reason": "InstallFailed",
        "status": "False",
        "type": "Released"
      }
    ],
    "failures": 12,
    "helmChart": "default/kubeapps-my-podinfo-5",
    "installFailures": 1,
    "lastAttemptedRevision": "6.0.0",
    "lastAttemptedValuesChecksum": "a619eeb520124dc343fb0222f3f7d0c27378a983",
    "observedGeneration": 2
  }
}]
```

Reply:
Yes, that case makes sense to me. My question was because it looked like above we already know that `installFailed == true` yet, we're on a conditional branch returning ready = true and installed successfully (i understand that condition may not be possible to get to true at that point).